CMakeLists.txt

```diff
@@ -401,9 +401,15 @@ set_directory_properties(PROPERTIES ADDITIONAL_MAKE_CLEAN_FILES "${CMAKE_SOURCE_
 
 # replace the default Debug flag of -g with -O0 -DDEBUG -ggdb3
 # this matches the flags of scons' debug build
-MESSAGE("Replacing flags used for Debug build ${OPT} -DDEBUG -ggdb3 ${EXTRA_FLAGS_CONFIG} ${EXTRA_FLAGS_DEBUG}")
-set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -DDEBUG -ggdb3 ${EXTRA_FLAGS_CONFIG} ${EXTRA_FLAGS_DEBUG}" CACHE STRING "change cmake's Debug flags to match scons' flags" FORCE)
-set(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -DDEBUG -ggdb3 ${EXTRA_FLAGS_CONFIG} ${EXTRA_FLAGS_DEBUG}" CACHE STRING "change cmake's Debug flags to match scons' flags" FORCE)
+if(HARDEN AND NOT WIN32)
+    MESSAGE("Replacing flags used for Debug build ${OPT} -Og -DDEBUG -ggdb3 ${EXTRA_FLAGS_CONFIG} ${EXTRA_FLAGS_DEBUG}")
+    set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -Og -DDEBUG -ggdb3 ${EXTRA_FLAGS_CONFIG} ${EXTRA_FLAGS_DEBUG}" CACHE STRING "change cmake's Debug flags to match scons' flags" FORCE)
+    set(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -Og -DDEBUG -ggdb3 ${EXTRA_FLAGS_CONFIG} ${EXTRA_FLAGS_DEBUG}" CACHE STRING "change cmake's Debug flags to match scons' flags" FORCE)
+else(HARDEN AND NOT WIN32)
```

Comment:
Given that hardening is enabled by default, setting the optimization to `-Og` when hardening is enabled also means that all debug builds will be compiled with `-Og` by default.  I don't know if that's desired, especially since when I brought up changing the `-O0` travis builds to `-Og` (to speed up the unit tests) with @jyrkive , he said it should remain at `-O0` there.

Reply:
You're the one who sent the suggestion about -Og.

I tried the OPT thing and it had no effect. It might be made to work but too much work.

I really am not in a place to fix this. I don't really know what hardening does (although I think I like the sound of it). Maybe the best solution is to simply write a HOWTO Compile document which says that if you want to compile a debug version, you need to either disable hardening or set -Og and just accept that every now and then we're going to get questioned about the #warning popping out of the runtime because someone didn't read the document.

My head hurts and I'm not in a place, mentally, to figure this out further. Do what you want and close this and the issue.

Reply:
Sorry about the confusion then - my question about `-Og` was really so I could know if I should disable _FORTIFY_SOURCE for both `-O0` and `-Og` or just `-O0`, I didn't mean to suggest changing the optimization level.  I'll implement the fix to disable _FORTIFY_SOURCE if the optimization level is `-O0` later today.

Reply:
It's also interesting that the issue seems to be happening only for Gregory. Like I said in #3518, I can make debug builds just fine.

Still, if the hardening causes as much as a *possibility* for builds to fail, disabling it for debug builds by default makes sense.